### PR TITLE
Add more data patterns, don't overwrite existing domains, dont offer blank patterns

### PR DESCRIPTION
**Pull request by jules2689**
## What this does
- Adds more options for dates
- Don't show an option if its empty or nil
- Don't offer to overwrite an existing domain in training

cc @richardwu

### CODE REVIEW

File: config/article_scrape_patterns.yml [hunk 1]
```diff
@@ -27,9 +27,36 @@ presets:
...
+      pattern: //meta[@namy='article_date_original']/@content
+    article_published_time: &article_published_time
       method: "xpath"
       pattern: "//meta[@property='article:published_time']/@content"
+    date: &date
+      method: xpath
+      pattern: //meta[@name='date']/@content
+    date_published: &date_published
+      method: xpath
+      pattern: //meta[@itemproy='datePublished']/@datetime
```
Comment (richardwu) [inline config/article_scrape_patterns.yml]: `itemproy` or `itemprop`?

File: config/article_scrape_patterns.yml [hunk 2]
```diff
@@ -27,9 +27,36 @@ presets:
...
+    og_published_time: &og_published_time
+      method: xpath
+      pattern: //meta[@property='og:published_time']/@content
+    original_publication_date: &original_publication_date
+      method: xpath
+      pattern: //meta[@namy='OriginalPublicationDate']/@content
```
Comment (richardwu) [inline config/article_scrape_patterns.yml]: `name`?

File: config/article_scrape_patterns.yml [hunk 3]
```diff
@@ -27,9 +27,36 @@ presets:
...
+    publication_date: &publication_date
+      method: xpath
+      pattern: //meta[@namy='publication_date']/@content
```
Comment (richardwu) [inline config/article_scrape_patterns.yml]: here too

File: config/article_scrape_patterns.yml [hunk 4]
```diff
@@ -27,9 +27,36 @@ presets:
...
+    publish_date: &publish_date
+      method: xpath
+      pattern: //meta[@namy='PublishDate']/@content
```
Comment (richardwu) [inline config/article_scrape_patterns.yml]: and here

File: config/article_scrape_patterns.yml [hunk 5]
```diff
@@ -27,9 +27,36 @@ presets:
...
+    rnews_datePublished: &rnews_datePublished
+      method: xpath
+      pattern: //meta[@property='rnews:datePublished']/@content
+    sailthru_date: &sailthru_date
+      method: xpath
+      pattern: //meta[@namy='sailthru.date']/@content
```
Comment (richardwu) [inline config/article_scrape_patterns.yml]: and here.

What's sailthru? Is it specific to 1 site? Does it need to be in presets?

File: config/article_scrape_patterns.yml [hunk 6]
```diff
@@ -27,9 +27,36 @@ presets:
...
+      pattern: //meta[@namy='OriginalPublicationDate']/@content
+    publication_date: &publication_date
+      method: xpath
+      pattern: //meta[@namy='publication_date']/@content
+    publish_date: &publish_date
+      method: xpath
+      pattern: //meta[@namy='PublishDate']/@content
+    rnews_datePublished: &rnews_datePublished
+      method: xpath
+      pattern: //meta[@property='rnews:datePublished']/@content
```
Comment (richardwu) [inline config/article_scrape_patterns.yml:63]: rnews also sounds like a domain-specific pattern and probably shouldn't go under presets

File: lib/news_scraper/trainer/uri_trainer.rb [hunk 7]
```diff
@@ -9,6 +9,8 @@ def initialize(uri)
       end
 
       def train
+        return if article_scrape_patterns['domains'].key?(@root_domain)
```
Comment (richardwu) [inline lib/news_scraper/trainer/uri_trainer.rb:12]: Nice catch

File: test/unit/trainers/uri_trainer_test.rb [hunk 8]
```diff
@@ -38,6 +38,15 @@ def test_train
           end
         end
       end
+
+      def test_train_on_trained_domain
```
Comment (richardwu) [inline test/unit/trainers/uri_trainer_test.rb]: I would rather this be descriptive of what behaviour it's testing for.

e.g. `test_train_on_trained_domain_returns_without_stepping_through_presets`

File: config/article_scrape_patterns.yml [hunk 9]
```diff
@@ -27,9 +27,36 @@ presets:
       method: "xpath"
       pattern: "//meta[@property='article:section']/@content"
   time:
-    published: &published_time
+    article_date_original: &article_date_original
+      method: xpath
+      pattern: //meta[@namy='article_date_original']/@content
```
Comment (richardwu) [inline config/article_scrape_patterns.yml]: here

File: config/article_scrape_patterns.yml (hunk 5 above)
Comment (jules2689) [reply config/article_scrape_patterns.yml]: Ive seen it on many sites, I'm going to research what it is, but it seems to be worth it

Comment (jules2689) [reply config/article_scrape_patterns.yml]: Sailthru seems to be a popular marketing platform: http://www.sailthru.com/
I've see it on a few sites as it seems to target article-based sites

File: config/article_scrape_patterns.yml (hunk 6 above)
Comment (jules2689) [reply config/article_scrape_patterns.yml:63]: rnews seems to be to support RDF https://www.w3.org/TR/rdf-sparql-query/

**Comment by jules2689** [comment]
FYI, for the contains stuff in the xpath:

```
irb(main):011:0> html1 = "<div class='class1 author class2'>hello</div>"
=> "<div class='class1 author class2'>hello</div>"
irb(main):013:0> noko_html_1 = Nokogiri::HTML(html1)
=> #<Nokogiri::HTML::Document:0x3fdb58cdb838 name="document" children=[#<Nokogiri::XML::DTD:0x3fdb58cdb004 name="html">, #<Nokogiri::XML::Element:0x3fdb58cda1b8 name="html" children=[#<Nokogiri::XML::Element:0x3fdb58cd3f5c name="body" children=[#<Nokogiri::XML::Element:0x3fdb58cd3bd8 name="div" attributes=[#<Nokogiri::XML::Attr:0x3fdb58cd3a0c name="class" value="class1 author class2">] children=[#<Nokogiri::XML::Text:0x3fdb58cd25a8 "hello">]>]>]>]>
irb(main):016:0> noko_html_1.xpath("//*[@class='author']")
=> []
irb(main):018:0> noko_html_1.xpath("//*[contains(@class, 'author')]")
=> [#<Nokogiri::XML::Element:0x3fdb58cd3bd8 name="div" attributes=[#<Nokogiri::XML::Attr:0x3fdb58cd3a0c name="class" value="class1 author class2">] children=[#<Nokogiri::XML::Text:0x3fdb58cd25a8 "hello">]>]
irb(main):020:0> noko_html_1.xpath("//*[contains(concat(' ', @class, ' '), ' author ')]")
=> [#<Nokogiri::XML::Element:0x3fdb58cd3bd8 name="div" attributes=[#<Nokogiri::XML::Attr:0x3fdb58cd3a0c name="class" value="class1 author class2">] children=[#<Nokogiri::XML::Text:0x3fdb58cd25a8 "hello">]>]
```

As you can see, the class=author doesnt work. the contains does, as does the concat/contains

```
irb(main):012:0> html2 = "<div class='class1 author-name class2'>hello</div>"
=> "<div class='class1 author-name class2'>hello</div>"
irb(main):014:0> noko_html_2 = Nokogiri::HTML(html2)
=> #<Nokogiri::HTML::Document:0x3fdb58d1ad94 name="document" children=[#<Nokogiri::XML::DTD:0x3fdb58d13648 name="html">, #<Nokogiri::XML::Element:0x3fdb58d12f18 name="html" children=[#<Nokogiri::XML::Element:0x3fdb58d12ae0 name="body" children=[#<Nokogiri::XML::Element:0x3fdb58d124dc name="div" attributes=[#<Nokogiri::XML::Attr:0x3fdb58d12400 name="class" value="class1 author-name class2">] children=[#<Nokogiri::XML::Text:0x3fdb58d0bc68 "hello">]>]>]>]>
irb(main):017:0> noko_html_2.xpath("//*[@class='author']")
=> []
irb(main):019:0> noko_html_2.xpath("//*[contains(@class, 'author')]")
=> [#<Nokogiri::XML::Element:0x3fdb58d124dc name="div" attributes=[#<Nokogiri::XML::Attr:0x3fdb58d12400 name="class" value="class1 author-name class2">] children=[#<Nokogiri::XML::Text:0x3fdb58d0bc68 "hello">]>]
irb(main):021:0> noko_html_2.xpath("//*[contains(concat(' ', @class, ' '), ' author ')]")
=> []
```

As you can see the class=author didnt work, the class contains author matched it when it should not have, and the concat/contains did not match it properly.

Therefore, concat/contains is needed

**Pull request merged**